### ISSUE TITLE
Add the option to instantiate the same rule more than once

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#6738]
This is a feature that we want to have for detekt 2.0

## Needed changes
There are multiple changes that should be made to make this possible. Here a list with all the changes that we need:

1. `RuleSet`s shouldn't instantiate the rules. They should provider factories to pass a confguration to create that rule.
    - **State:** #6733
1. `Rule`s shouldn't pick their own configuration from its parent configuration. We should provide their configuration directly.
    - **State:** #6767
1. The `Rule` id can't be hardcoded. If a rule can have multiple instances the id should change between those instances.
    - **State:** No started
1. The `Config` class should provide an API to read all the subconfigurations inside it.
    - **State:** #6842
1. The core should read the `Config` and instantiate the rules from there.
    - **State:** This should be the easiest. We need to align on the new yaml key to find the rule but the change should be as easy as this:
```kotlin
.flatMap { (ruleSet, ruleSetConfig) ->
    ruleSetConfig.subconfigs
        .map { (configId, config) ->
            val ruleId = config.valueOrNull("rule") ?: configId
            val ruleProvider = checkNotNull(ruleSet.rules[ruleId]) {
                "Unknown rule ${ruleSet.id}:$ruleId"
            }
            ruleProvider.invoke(config)
        }
}
```

----

This is a huge change. There is code that assumes that the rules have a hardcoded id or that the there is only one instance of each. And this is a breaking change, we need to change the API for 3 main classes/interfaces: `RuleSet`, `Rule` and `Config`. This will impact every single third party rule out there.

Also, those are the big "chunks" that should be done, then a lot of minor changes should be made on the way to be able to make those possible.

**Comment by BraisGabin** [comment]
You are correct, that's exactly the use case. Could I ask you how do you imagine the configuration should look like to make your use case work?

**Comment by BraisGabin** [comment]
I need to write down all my thoughts about Rule/Issue/Finding and all that part of the detekt API:

We need to talk about rule ids. Right now it's easy! Every `Rule` has an `Issue`. The `Issue` stores the `id` and the `description` of the `Rule`. And `Rule.id` is just syntax sugar for `Rule.issue.id`.

That works when you only have one instance per rule. But once you have multiple instance of the same rule the thing gets messy really quick.

First of all the rules need 2 `id`s:
- One to identify the type of the rule (needed to link for its documentation, for example)
- One to identify the instance of the rule (as a user I want to know which configuration of the rule raised that issue)

The first one can be hardcode. It can even be the `simpleName` or the `fullQualifiedName` of the `Rule`. But the second is way trickier. First of all, that id should come from `Config` (aka yaml). And how are we going to expose those different ids?

`Issue` is a nice API. It's used in two different places: the `Rule` and the `Finding`. `Issue` is like a "RuleInfo" class. It is an elegant way to tell the `Finding` who created it but without having a reference to a `Rule`.

So, if we need two different ids for a rule both should be inside `Issue`.

An issue right now is:

```kotlin
class Issue(
    val id: String,
    val description: String,
)
```

I would propouse:

```kotlin
class Issue(
    val id: String,
    val ruleId: String,
    val description: String,
)
```

Where `instanceId`, in general, would be the same as `ruleId`. It will change only when we define a different id.

So, how would this look like on the yaml?

```yaml
style:
  ForbiddenJavaIoFile:
    active: true
    rule: ForbiddenImport
    ignoreAnnotated:
      - "JavaIoFileIsOk"
    imports:
      - value: "java.io.File"
        reason: 'case-sensitivity issues'
  ForbiddenImport:
    active: true
    imports:
      - value: "some.unrelated.import"
        reason: "some other reason"
```

A configuration like that will create this two instances like this:

```kotlin
listOf(
  ForbiddenImport(
    Issue(id = "ForbiddenJavaIoFile", ruleId = "ForbiddenImport", ...),
    ...
  ),
  ForbiddenImport(
    Issue(id = "ForbiddenImport", ruleId = "ForbiddenImport", ...),
    ...
  ),
)
```

Also I would vote to use `value class` for the `id`s to have better type safety. If we have 2 different ids it will be difficult to know which oen use.

**Comment by marschwar** [comment]
Sounds good. I like the proposal.

**Comment by marschwar** [comment]
> I have to admit, I don't know if there is already a built-in feature in YAML that would solve this issue. If there is not though, maybe we would want to provide our own "templates" or other inheritance mechanism in the configuration files.

There is the concept of anchors that you might be able to use.

**Comment by VovaStelmashchuk** [comment]
My proposal about implement config in case user need a the rule twice or more time. We add special char for separate rule name and the rule id. 
```yaml
style:
  MaxLineLength:
    maxLineLength: 100
  MaxLineLength%some_id:
    maxLineLength: 100
```
On parsing stage we can check the key for contains % and split the string to rule name. Then add two rule to ruleset.
- MaxLineLenght()
- MaxLineLenght(id = some_id)

I use a `%` as spec char.

Is the proposal have any sence ?

**Comment by BraisGabin** [comment]
It does have sense. I like it! I would say that `:` is a good candidate `MaxLineLength:some_id`

```yaml
style:
  MaxLineLength:
    maxLineLength: 100
  MaxLineLength:some_id:
    maxLineLength: 100
```

The only problem with that is that `yaml` uses `:` a lot already so I don't know if it is a good idea or not.

Other char options: `;`, `/`, `|`, `=`, `,` and `-`.

**Comment by VovaStelmashchuk** [comment]
I suggest a `+` as special char. The `+` is already using as splitter for alias in email naming convention. Also, the plus symbol clearly indicates an addition/extension.

[Email Sub-addressing](https://en.wikipedia.org/wiki/Email_address#Sub-addressing)

**Comment by VovaStelmashchuk** [comment]
I don't agree with the @ special character, this character is used in so many programs for communication as a special character, (slack, discord, github, mail...)

This will be inconvenient for people when communicating about their detekt rules. Each time you type the @ some program suggest a popup.

**Comment by VovaStelmashchuk** [comment]
Ok, I think we need other options in discussion:)

One more point about works with ids, Do the [special char] must the valid id itself?

I think it should, because otherwise we will need extra validation of yaml parsing.

**Comment by marschwar** [comment]
Do we know if anyone is using yaml schema validation for the detekt config? I think auto-complete should still work for the general rule instance, but validation would fail. But I guess that is already the case for 3rd party rules.

We should take into account #4162, when thinking about this.

**Comment by BraisGabin** [comment]
`/` can work. We can think of the ID as a path or the name of the rule as a directory. I like the discussion to see pros and cons

**Comment by BraisGabin** [comment]
Ok, at #7263 I have the first working version of detekt running multiple instances of the same rule with different configurations.

There is still work to do but it works!

Missing things:
- tests, lots of them.
- Rework the config validator, our config files now are different! Probably all this part needs a full refactor. Now that we have the annotations on the rules to generate the config file we could also autogenerate the code that validates it... I'm not sure but it would be cool. But I don't want to go that far for this issue.
- Documentation, we need to explain how to use this feature.
- Take some decissions about naming and how do we want the configuration to look like. For now I'm using `/` as the separator but I would like to hear feedback about how it looks like.
- Work on the reports and baseline. We need to use the instance id to report the issues, no the name of the rule itself.
